File: ruby/binary_search/lib/binary_search.rb
require_relative File.join(App.root, 'lib', 'read_json_file.rb')

INPUT_JSON = File.join(App.root, 'data', 'input.json')

class BinarySearch
  attr_reader :input, :no_to_search, :iterations
  include ReadJsonFile


  def initialize(number_to_search)
    @input = read_input_file(INPUT_JSON)['input'].sort
    @no_to_search = number_to_search.to_i
  end

  def search(low_index=0, high_index=@input.length - 1, iterations=0, input=@input)
    @iterations = iterations + 1
    if high_index >= low_index
      mid_index = low_index + ((high_index - low_index)/2.0).ceil
      # puts "#{@iterations} - #{low_index}, #{mid_index} and #{high_index} of #{@input}"
      # puts "Comparing #{input[mid_index]} with #{no_to_search}"
      if input[mid_index] == no_to_search
        return mid_index
      elsif input[mid_index] < no_to_search
        search(mid_index + 1, high_index, @iterations)
      else input[mid_index] > no_to_search
        search(low_index, mid_index - 1, @iterations)
      end
    else
      return -1
    end
  end

  def self.valid_number?(number)
    number.match?(/^\d+$/)
  end
end

File: python/algorithms/binary_search.py
import math
import pdb


class BinarySearch:
    def __init__(self, search_array):
        search_array.sort()
        self.search_array = search_array
        self.iteration = 0

    def search(self, target, low_index=0, high_index=None):
        if high_index is None:
            high_index = high_index or len(self.search_array) - 1
        self.iteration += 1
        if high_index >= low_index:
            mid_index = low_index + math.ceil((high_index - low_index)/2.0)
            if self.search_array[mid_index] == target:
                return mid_index
            elif self.search_array[mid_index] < target:
                return self.search(target, mid_index + 1, high_index)
            elif self.search_array[mid_index] > target:
                return self.search(target, low_index, mid_index - 1)
        else:
            return -1
